backend/czi_hosted/data_common/cache.py

```diff
@@ -0,0 +1,219 @@
+                return True
+        else:
+            return False
+
+
+class CacheItemInfo(object):
+    """
+    This class stores a pointer to and metadata about the cached item. When the data is accessd the
+    update function is called to track the latest access
+    """
+
+    def __init__(self, cache_item: CacheItem(), timestamp):
+        self.cache_item = cache_item
+        self.last_access = timestamp
+        self.num_access = 1
+
+    def update_latest_cache_access(self):
+        self.last_access = time.time()
+        self.num_access += 1
+
+
+class CacheManager(object):
+    """An base class to manage cached data. This is intended to be used as a context manager
+    for handling api requests.  When the context is created, the data_adaptor is either loaded or
+    retrieved from a cache.  In either case, the reader lock is taken during this time, and release
+    when the context ends.  This class currently implements a simple least recently used cache,
+    which can delete data from the cache to make room for a new one.
+
+    This is the intended usage pattern:
+
+           cache_manager = DataCacheManager(max_cached=..., timelimmit_s = ...)
+           with cache_manager.data_adaptor(location, app_config) as data_adaptor:
+               # use the data_adaptor for some operation
+    """
+
+    # FIXME:   If the number of active datasets exceeds the max_cached, then each request could
+    # lead to a dataset being deleted and a new only being opened: the cache will get thrashed.
+    # In this case, we may need to send back a 503 (Server Unavailable), or some other error message.
+
+    # NOTE:  If the actual dataset is changed.  E.g. a new set of datafiles replaces an existing set, or the location
+    # of the data is updated then the cache will not react to this, however once the cache time limit is reached,
+    # the data will automatically be refreshed.
+
+    def __init__(self, max_cached: int, timelimit_s: int = None):
+        self.data = {}
+
+        # lock to protect the datasets
+        self.lock = threading.Lock()
+
+        #  The number of items to cache.  When max_cached is reached, the least recently used
+        #  cache is replaced with the newly requested one.
+        #  TODO:  This is very simple.  When this class is used to cache the actual datasets
+        #   this can be improved by taking into account how much space is actually taken by each dataset,
+        #   instead of arbitrarily picking a max number to cache.
+        self.max_cached = max_cached
+
+        # items are automatically removed from the cache once this time limit (in seconds) is reached
+        self.timelimit_s = timelimit_s
+
+    @contextmanager
+    def get(self, cache_key: str, create_data_function: Optional[Callable] = None,
+            create_data_args: object = {}):
+        """"
+        If no data is stored under the given cache_key, pass the create_data_lamba to the CacheItem.get method
```

Comment:
Could you write a better pydoc here? I didn't quite get the gist of the function with the above comment. Also `create_data_lamba` doesn't exist? Should this be `create_data_function`? And why was it important to call out?